File: wx_sdk/ios/Classes/WxSdkPlugin.m
#import "WxSdkPlugin.h"
#import <CJBase/CJBase.h>
#import <NIMKit/NIMKit.h>
#import "FlutterBoost.h"
#import "NimSdkUtilPlugin.h"

static NSString *wxSDKResultKey = @"flutter_result";

@implementation WxSdkPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"wx_sdk"
            binaryMessenger:[registrar messenger]];
  WxSdkPlugin* instance = [[WxSdkPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([@"getPlatformVersion" isEqualToString:call.method]) {
    result([@"iOS " stringByAppendingString:[[UIDevice currentDevice] systemVersion]]);
  } else {
      SEL sel = NSSelectorFromString(call.method);
      if([WxSdkPlugin respondsToSelector:sel]) {
          NSDictionary *params = call.arguments;
          NSMutableDictionary *p = params?params.mutableCopy : @{}.mutableCopy;
          [p setObject:result forKey:wxSDKResultKey];
          [WxSdkPlugin performSelector:sel
                            withObject:p
                            afterDelay:0];
      }else {
          result(FlutterMethodNotImplemented);
      }
  }
}

+ (void)wxlogin
{
    if ([WXApi isWXAppInstalled]) {
        SendAuthReq* req = [[SendAuthReq alloc] init];
        req.scope = @"snsapi_userinfo";
        req.state = @"get_access_token";
        [WXApi sendReq:req];
    }
}

/// 分享到微信
/// web:12
+ (void)share:(NSDictionary *)params
{
    WXMediaMessage *message = [WXMediaMessage message];
    NSString *title = params[@"title"];
    NSString *content = params[@"content"];
    NSString *url = params[@"url"];
    NSNumber *type = params[@"type"];
    if (title == nil) {
        message.title = @"";
    }
    else
    {
        message.title = title;
    }
    message.description     = content;
    message.messageExt      = content;
    message.messageAction   = content;
    
    WXWebpageObject *ext = [WXWebpageObject object];
    ext.webpageUrl = url;
    
    message.mediaObject = ext;
    
    SendMessageToWXReq* req = [[SendMessageToWXReq alloc] init];
    req.bText = NO;
    req.message = message;
    req.scene = WXSceneSession;
    
    [WXApi sendReq:req];
}

- (void)onResp:(BaseResp*)resp{
    // 分享成功是否的监听
    if([resp isKindOfClass:[SendMessageToWXResp class]])
    {
        if(resp.errCode == 0){
            
        }
    }
    // 授权结果监听
    else if([resp isKindOfClass:[SendAuthResp class]])
    {
        if (resp.errCode == 0) {
            SendAuthResp *_resp = (SendAuthResp*)resp;
            NSString* accessToken =  _resp.code;
            ZZLog(@"WetChat AccessToken =======> %@", accessToken);
            // 拿到TOKEN后去服务端认证下
            if ([_resp.state isEqualToString:@"get_access_token_bind"]) {
                [WxSdkPlugin wxBindCode:accessToken];
            }else{
                // 拿到TOKEN后去服务端认证下
                [WxSdkPlugin sendLoginAuth:accessToken
                                    result:^(BaseModel *model)
                {
                    [self onWxLoginResp:model code:accessToken];
                }];
            }
        }
        else
        {
            [UIViewController showError:@"用户取消或者拒绝了微信授权登录"];
        }
    }
}

- (void)onWxLoginResp:(BaseModel *)model code:(NSString *)code
{
    if (model.success)
    {
        // 登录
        [NimSdkUtilPlugin doLogin:model.data];
        
    }else if ([model.error isEqualToString:@"1"]){
        [UIViewController hideHUD];
        // 显示绑定手机页面
        NSDictionary* data = model.data;
        if (data != nil) {
            NSString *union_id  = [data objectForKey:@"union_id"];
            NSString *headimg   = [data objectForKey:@"head_img"];
            NSString *nick_name = [data objectForKey:@"nick_name"];
            
            [FlutterBoostPlugin open:@"phone_bind" urlParams:@{
                @"union_id": union_id?:[NSNull null],
                @"headimg": headimg?:[NSNull null],
                @"nick_name": nick_name?:[NSNull null],
                @"code": code?:[NSNull null]
            } exts:@{@"animated": @(YES)} onPageFinished:^(NSDictionary *d) {
                
            } completion:^(BOOL success) {
                
            }];
        }
    }else if([model.error isEqualToString:@"8"])
    {
        [UIViewController hideHUD];
        // 账号冻结了
        //        [LoginTaskManager unfreezeAccount];
    }
    else{
        [UIViewController hideHUD];
        [UIViewController showError:model.errmsg];
    }
}

+ (void)wxBindCode:(NSString *)code
{
    NSString *url =  [NSString stringWithFormat:@"%@/g2/user/wx/bind",kBaseUrl];
    NSString *accid = [[[NIMSDK sharedSDK] loginManager] currentAccount];
    NSDictionary *params = @{@"accid":accid ? : @"",
                             @"code":code ? : @"",
                             @"union_id":@"",
                             @"app_key":@"wx0f56e7c5e6daa01a"};
    [UIViewController showWaiting];
    
    co_launch(^{
        BaseModel *model = await([HttpHelper post:url params:params]);
        
        [UIViewController hideHUD];
        if(co_getError()) {
            [UIViewController showError:CJ_net_err_msg];
        }else if(model.success) {
            [UIViewController showSuccess:@"绑定成功"];
        }else {
            [UIViewController showError:model.errmsg];
        }
    });
}

+ (void)sendLoginAuth:(NSString*)accessToken
               result:(void (^)(BaseModel *model))result
{
    ZZLog(@"sendLoginAuth  accessToken  %@", accessToken);
    //加入参数
    NSMutableDictionary *params = [NSMutableDictionary new];
    [params setValue:accessToken forKey:@"code"];
    [params setValue:@"wx0f56e7c5e6daa01a" forKey:@"app_key"];
    [UIViewController showLoadingWithMessage:@"登录中..."];
    
    co_launch(^{
        BaseModel *model = await([HttpHelper post:kWechatLoginUrl params:params]);
        
        [UIViewController hideHUD];
        if(co_getError()) {
            [UIViewController showError:CJ_net_err_msg];
        }else if(result) {
            result(model);
        }
    });
}

/// 查询微信绑定状态
+ (void)wxBindStatus:(NSDictionary *)params
{
    FlutterResult result = params[wxSDKResultKey];
    NSString *url =  [NSString stringWithFormat:@"%@/g2/user/wx/bind/exist",kBaseUrl];
    NSString *accid = [[[NIMSDK sharedSDK] loginManager] currentAccount];
    NSDictionary *p = @{@"accid":accid ? : @"",
                         @"app_key":@"wx0f56e7c5e6daa01a"};
    
    co_launch(^{
        BaseModel *model = await([HttpHelper post:url params:p]);
        
        if(co_getError()) {
            result(@(NO));
            [UIViewController showError:CJ_net_err_msg];
        }else {
            if (model.success && !cj_nil_object(model.data)) {
                result(@(YES));
            }else{
                result(@(NO));
            }
        }
    });
}

/// 解绑微信
+ (void)unBindWeChat:(NSDictionary *)params
{
    FlutterResult result = params[wxSDKResultKey];
    NSString *url =  [NSString stringWithFormat:@"%@/g2/user/wx/untying",kBaseUrl];
    NSString *accid = [[[NIMSDK sharedSDK] loginManager] currentAccount];
    NSDictionary *p = @{@"accid":accid ? : @"",
                             @"app_key":@"wx0f56e7c5e6daa01a"};
    
    co_launch(^{
        BaseModel *model = await([HttpHelper post:url params:p]);
        if(co_getError()) {
            result(@(NO));
            [UIViewController showError:CJ_net_err_msg];
        }else {
            if (model.success) {
                result(@(YES));
                [UIViewController showSuccess:@"解绑成功"];
            }else{
                result(@(NO));
                [UIViewController showError:model.errmsg];
            }
        }
    });
}

@end

File: wx_sdk/android/src/main/java/com/cajian/wx_sdk/WxSdkPlugin.kt
package com.cajian.wx_sdk

import android.content.Context
import android.os.Build
import android.util.Log
import android.widget.Toast
import com.blankj.utilcode.util.LogUtils
import com.blankj.utilcode.util.SPUtils
import com.blankj.utilcode.util.ToastUtils
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import com.netease.nim.uikit.api.NimUIKit
import com.netease.nimlib.sdk.NIMSDK
import com.netease.nimlib.sdk.RequestCallback
import com.netease.nimlib.sdk.auth.LoginInfo
import com.tencent.mm.opensdk.modelbase.BaseResp
import com.tencent.mm.opensdk.modelmsg.SendAuth
import com.tencent.mm.opensdk.modelmsg.SendMessageToWX
import com.tencent.mm.opensdk.modelmsg.WXMediaMessage
import com.tencent.mm.opensdk.modelmsg.WXWebpageObject
import com.tencent.mm.opensdk.openapi.IWXAPI
import com.tencent.mm.opensdk.openapi.WXAPIFactory
import com.youxi.chat.base.Config
import com.youxi.chat.base.PopupManager
import com.youxi.chat.base.net.BaseModel
import com.youxi.chat.base.net.CommonApi
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.PluginRegistry.Registrar
import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import java.lang.reflect.InvocationTargetException

/** WxSdkPlugin  */
class WxSdkPlugin private constructor(private val mRegistrar: Registrar) : MethodCallHandler {

    private val mContext: Context
    private val mWxApi: IWXAPI

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        // 适配iOS方法名,iOS中有参数的方法名带有冒号,Android中需要去掉
        val methodName = call.method.replace(":", "")
        if ("getPlatformVersion" == methodName) {
            result.success("Android " + Build.VERSION.RELEASE)
            return
        }
        try {
            val params = if (call.arguments == null) mapOf<Any, Any?>() else call.arguments
            val method = javaClass.getDeclaredMethod(methodName, Map::class.java, MethodChannel.Result::class.java)
            method.isAccessible = true
            method.invoke(this, params, result)
            return
        } catch (e: NoSuchMethodException) {
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        } catch (e: InvocationTargetException) {
            e.printStackTrace()
        } catch (e: Exception) {
            e.printStackTrace()
        }

        ToastUtils.showLong("method ${call.method} notImplemented")
        result.notImplemented()
    }

    /**
     * 微信登录
     */
    private fun wxlogin(params: Map<*, *>, result: MethodChannel.Result) {
        if (mWxApi.isWXAppInstalled) {
            val req = SendAuth.Req()
            req.scope = "snsapi_userinfo"
            req.state = "get_access_token"
            mWxApi.sendReq(req)
        } else {
            Toast.makeText(mContext, "您还没有安装微信,请下载安装", Toast.LENGTH_SHORT).show()
        }
        result.success(null)
    }

    /**
     * 微信分享
     */
    private fun share(params: Map<*, *>, result: MethodChannel.Result) {
        val title = params["title"] as String?
        val content = params["content"] as String
        val url = params["url"] as String
        val type = params["type"] as Int

        val message = WXMediaMessage()
        message.title = title ?: ""
        message.description = content
        message.messageExt = content
        message.messageAction = content

        val ext = WXWebpageObject()
        ext.webpageUrl = url

        message.mediaObject = ext

        val req = SendMessageToWX.Req()
        req.transaction = buildTransaction("text")
        req.message = message
        req.scene = SendMessageToWX.Req.WXSceneSession

        mWxApi.sendReq(req)
    }

    private fun onResp(_resp: BaseResp) {
        if (_resp is SendMessageToWX.Resp) {
            // 分享成功是否的监听
            if (_resp.errCode == 0) {
                // 什么也不做
            }
        } else if (_resp is SendAuth.Resp) {
            // 授权结果监听
            if (_resp.errCode == 0) {
                val resp = _resp
                val accessToken = resp.code
                LogUtils.d("Wechat AccessToken = $accessToken")
                if ("get_access_token_bind" == resp.state) {
                    wxBindCode(accessToken)
                } else { // 拿到TOKEN后去服务端认证下
                    sendLoginAuth(accessToken)
                }
            }
        } else {
            Toast.makeText(mContext, "用户取消或者拒绝了微信授权登录", Toast.LENGTH_SHORT).show()
        }
    }

    private fun onWxLoginResp(model: BaseModel<LoginInfoResp>, code: String) {
        if (model.success()) {
            val loginInfo = LoginInfo(model.data.accid, model.data.token)
            NIMSDK.getAuthService()
                    .login(loginInfo)
                    .setCallback(object : RequestCallback<Any?> {
                        override fun onSuccess(param: Any?) {
                            ToastUtils.showShort("登录成功")
                            stashLoginInfo(model.data.accid, model.data.token)
                        }

                        override fun onFailed(code: Int) {
                            ToastUtils.showShort("登录失败, 错误码=$code")
                        }

                        override fun onException(exception: Throwable?) {
                            ToastUtils.showShort("登录异常, 异常原因=${exception?.message}")
                        }
                    })
        } else if ("1" == model.error) {
            // TODO 绑定手机
        } else if ("8" == model.error) {
            // TODO 账号冻结
        } else {
            ToastUtils.showShort(model.errmsg)
        }
    }

    private fun stashLoginInfo(accid: String, token: String) {
        SPUtils.getInstance().put("flutter.accid", accid)
        SPUtils.getInstance().put("flutter.token", token)
    }

    private fun wxBindCode(code: String) {
        val accid = NimUIKit.getAccount()
        val json = mapOf(
                "accid" to accid,
                "code" to code,
                "union_id" to "",
                "app_key" to APP_ID
        )

        CommonApi.getApi().postJson(Config.wechatBindUrl, json = json)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : Observer<String> {
                    override fun onComplete() {
                        PopupManager.hideLoading(mContext)
                    }

                    override fun onSubscribe(d: Disposable) {
                        PopupManager.showLoading(mContext)
                    }

                    override fun onNext(t: String) {
                        val model = GsonBuilder().setLenient().create().fromJson(t, BaseModel::class.java)
                        if (model.success()) {
                            ToastUtils.showShort("绑定成功")
                        } else {
                            ToastUtils.showShort(model.errmsg)
                        }
                    }

                    override fun onError(e: Throwable) {
                        PopupManager.hideLoading(mContext)
                        ToastUtils.showShort(e.message)
                    }
                })
    }

    private fun sendLoginAuth(accessToken: String) {
        LogUtils.d("sendLoginAuth accessToken = $accessToken")

        val json = mapOf(
                "code" to accessToken,
                "app_key" to APP_ID
        )

        CommonApi.getApi().postJson(Config.wechatLoginUrl, json = json)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : Observer<String> {
                    override fun onComplete() {
                        PopupManager.hideLoading(mContext)
                    }

                    override fun onSubscribe(d: Disposable) {
                        PopupManager.showLoading(mContext, loadingTips = "登录中...")
                    }

                    override fun onNext(t: String) {
                        val model = GsonBuilder().setLenient().create()
                                .fromJson<BaseModel<LoginInfoResp>>(t, object : TypeToken<BaseModel<LoginInfoResp>>() {}.type)
                        onWxLoginResp(model, accessToken)
                    }

                    override fun onError(e: Throwable) {
                        PopupManager.hideLoading(mContext)
                        ToastUtils.showShort(e.message)
                    }
                })
    }

    /**
     * 查询微信绑定状态
     */
    private fun wxBindStatus(params: Map<*, *>, result: MethodChannel.Result) {
        val accid = NimUIKit.getAccount()

        val json = mapOf(
                "accid" to accid,
                "app_key" to APP_ID
        )

        CommonApi.getApi().postJson(Config.wechatStatusUrl, json = json)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : Observer<String> {
                    override fun onComplete() {
                        PopupManager.hideLoading(mContext)
                    }

                    override fun onSubscribe(d: Disposable) {
                        PopupManager.showLoading(mContext)
                    }

                    override fun onNext(t: String) {
                        val model = GsonBuilder().setLenient().create().fromJson(t, BaseModel::class.java)
                        if (model.success()) {
                            result.success(true)
                        } else {
                            result.success(false)
                        }
                    }

                    override fun onError(e: Throwable) {
                        PopupManager.hideLoading(mContext)
                        result.success(false)
                    }
                })
    }

    /**
     * 解绑微信
     */
    private fun unBindWeChat(params: Map<*, *>, result: MethodChannel.Result) {
        val accid = NimUIKit.getAccount()

        val json = mapOf(
                "accid" to accid,
                "app_key" to APP_ID
        )

        CommonApi.getApi().postJson(Config.wechatUnbindUrl, json = json)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : Observer<String> {
                    override fun onComplete() {
                        PopupManager.hideLoading(mContext)
                    }

                    override fun onSubscribe(d: Disposable) {
                        PopupManager.showLoading(mContext, loadingTips = "登录中...")
                    }

                    override fun onNext(t: String) {
                        val model = GsonBuilder().setLenient().create().fromJson(t, BaseModel::class.java)
                        if (model.success()) {
                            ToastUtils.showShort("解绑成功")
                        } else {
                            ToastUtils.showShort(model.errmsg)
                        }
                    }

                    override fun onError(e: Throwable) {
                        PopupManager.hideLoading(mContext)
                        ToastUtils.showShort(e.message)
                    }
                })
    }

    private fun buildTransaction(type: String?): String {
        return if (type == null) System.currentTimeMillis().toString() else type + System.currentTimeMillis();
    }

    companion object {
        private const val APP_ID = "wxa590aacb6f7b637b"
        private val TAG = WxSdkPlugin::class.java.simpleName
        /** Plugin registration.  */
        @JvmStatic
        fun registerWith(registrar: Registrar) {
            val channel = MethodChannel(registrar.messenger(), "wx_sdk")
            channel.setMethodCallHandler(WxSdkPlugin(registrar))
            Log.d(TAG, "registerWith")
        }
    }

    init {
        mContext = mRegistrar.context()
        mWxApi = WXAPIFactory.createWXAPI(mContext, APP_ID)
        mWxApi.registerApp(APP_ID)
    }

    data class LoginInfoResp(
            val accid: String,
            val token: String
    )
}